R/death_functions.R

```diff
@@ -960,6 +1000,503 @@ death_injury_matrix_count <- function(ph.data,
+    if (contributing_logic == "ANY") {
+      # ANY logic - at least one contributing code must be present
+      ph.data[meets_criteria == TRUE,
+              meets_criteria := Reduce(`|`, lapply(.SD, function(x) x %in% contributing_codes)),
+              .SDcols = contrib_col_names]
+    }
+
+    if (contributing_logic == "ALL") {
+      # For each contributing code, check if it appears in any contributing columns
+      # If it does, it continues to `meet_criteria`.
+      # If it doesn't, it is removed from further comparisons
+      for(code in contributing_codes) {
+        code_present <- ph.data[meets_criteria == TRUE,
+                                Reduce(`|`, lapply(.SD, function(x) x %in% code)),
+                                .SDcols = contrib_col_names]
+        ph.data[meets_criteria == TRUE, meets_criteria := meets_criteria & code_present]
+      }
+    }
+
+  # Calculate YPLL if needed ----
+    if (!is.null(ypll_age)) {
+      ph.data[, (ypll_col_name) := fifelse(get(death_age_col) < ypll_age,
+                                           ypll_age - get(death_age_col),
+                                           0)]
+    }
+
+  # Create summary of death and ypll counts ----
+    if (is.null(ypll_age)) {
+      # All causes
+      all_causes <- ph.data[, list(cause.of.death = "All causes",
+                                deaths = .N), by = group_by]
+
+      # Specific multicause
+      specific_cause <- ph.data[meets_criteria == TRUE,
+                                list(cause.of.death = cause_label,
+                                  deaths = .N), by = group_by]
+    } else {
+      # All causes with YPLL
+      all_causes <- ph.data[, list(cause.of.death = "All causes",
+                                deaths = .N,
+                                temp_ypll = sum(get(ypll_col_name), na.rm = TRUE)),
+                            by = group_by]
+
+      # Specific multicause with YPLL
+      specific_cause <- ph.data[meets_criteria == TRUE,
+                                list(cause.of.death = cause_label,
+                                  deaths = .N,
+                                  temp_ypll = sum(get(ypll_col_name), na.rm = TRUE)),
+                                by = group_by]
+
+      # Rename YPLL column
+      setnames(all_causes, "temp_ypll", ypll_col_name)
+      setnames(specific_cause, "temp_ypll", ypll_col_name)
+    }
+
+    # Combine results
+    result <- rbind(all_causes, specific_cause)
+
+  # Create rows for zero values ----
+    # Select columns to use for combination
+    cols_to_use <- setdiff(names(result), c('deaths', grep('^ypll_', names(result), value = TRUE)))
+
+    # Create list of unique values for each selected column
+    unique_col_vals <- lapply(cols_to_use, function(col) unique(result[[col]]))
+    names(unique_col_vals) <- cols_to_use
+
+    # Use CJ to create all combinations
+    template_xyz <- do.call(data.table::CJ, unique_col_vals)
```

Comment:
Why is this necessary? It seems like a potentially expensive operation when you could use the absence of something to indicate 0.